x-pack/plugins/fleet/server/services/secrets.ts

```diff
@@ -102,47 +105,136 @@ export async function createSecrets(opts: {
   }
 }
 
-export async function deleteSecret(opts: {
+export async function deleteSecretsIfNotReferenced(opts: {
   esClient: ElasticsearchClient;
-  id: string;
-}): Promise<DeleteResponse['result']> {
-  const { esClient, id } = opts;
-  let res: DeleteResponse;
+  soClient: SavedObjectsClientContract;
+  ids: string[];
+}): Promise<void> {
+  const { esClient, soClient, ids } = opts;
+  const logger = appContextService.getLogger();
+  const packagePoliciesUsingSecrets = await findPackagePoliciesUsingSecrets({
+    soClient,
+    ids,
+  });
+
+  if (packagePoliciesUsingSecrets.length) {
+    packagePoliciesUsingSecrets.forEach(({ id, policyIds }) => {
+      logger.debug(
+        `Not deleting secret with id ${id} is still in use by package policies: ${policyIds.join(
+          ', '
+        )}`
+      );
+    });
+  }
+
+  const secretsToDelete = ids.filter((id) => {
+    return !packagePoliciesUsingSecrets.some((packagePolicy) => packagePolicy.id === id);
+  });
+
+  if (!secretsToDelete.length) {
+    return;
+  }
+
+  await _deleteSecrets({
+    esClient,
+    ids: secretsToDelete,
+  });
+}
+
+export async function findPackagePoliciesUsingSecrets(opts: {
+  soClient: SavedObjectsClientContract;
+  ids: string[];
+}): Promise<Array<{ id: string; policyIds: string[] }>> {
+  const { soClient, ids } = opts;
+  const packagePolicies = await packagePolicyService.list(soClient, {
+    kuery: `ingest-package-policies.secret_references.id: (${ids.join(' or ')})`,
+    perPage: SO_SEARCH_LIMIT,
+    page: 1,
+  });
+
+  if (!packagePolicies.total) {
+    return [];
+  }
+
+  // create a map of secret_references.id to package policy id
+  const packagePoliciesBySecretId = packagePolicies.items.reduce((acc, packagePolicy) => {
+    packagePolicy?.secret_references?.forEach((secretReference) => {
+      if (!acc[secretReference.id]) {
+        acc[secretReference.id] = [];
+      }
+      acc[secretReference.id].push(packagePolicy.id);
+    });
+    return acc;
+  }, {} as Record<string, string[]>);
+
+  const res = [];
+
+  for (const id of ids) {
+    if (packagePoliciesBySecretId[id]) {
+      res.push({
+        id,
+        policyIds: packagePoliciesBySecretId[id],
+      });
+    }
+  }
+
+  return res;
+}
+
+// use deleteSecretsIfNotReferenced
```

Comment:
nit: is this comment necessary? it should be "used by deleteSecretsIfNotReferenced"